File: packages/core/tests/unit/derived-roles/validator.test.ts
import { describe, it, expect } from 'vitest';
import { DerivedRolesValidator } from '../../../src/derived-roles/validator';
import { CircularDependencyError } from '../../../src/derived-roles/types';
import type { ValidatedDerivedRolesPolicy } from '../../../src/policy/schema';

describe('DerivedRolesValidator', () => {
  let validator: DerivedRolesValidator;

  beforeEach(() => {
    validator = new DerivedRolesValidator();
  });

  describe('Schema Validation', () => {
    it('should accept valid derived roles policy', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'valid-policy' },
        spec: {
          definitions: [
            {
              name: 'owner',
              parentRoles: ['user'],
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).not.toThrow();
    });

    it('should reject duplicate role names', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'duplicate-policy' },
        spec: {
          definitions: [
            {
              name: 'owner',
              parentRoles: ['user'],
              condition: { expression: 'true' },
            },
            {
              name: 'owner', // Duplicate
              parentRoles: ['admin'],
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).toThrow('Duplicate derived role name: owner');
    });

    it('should reject invalid role names with special characters', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'invalid-name-policy' },
        spec: {
          definitions: [
            {
              name: 'owner@role', // Invalid character
              parentRoles: ['user'],
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).toThrow('Invalid derived role name');
    });

    it('should accept role names with hyphens and underscores', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'valid-names' },
        spec: {
          definitions: [
            {
              name: 'owner-role',
              parentRoles: ['user'],
              condition: { expression: 'true' },
            },
            {
              name: 'admin_role',
              parentRoles: ['user'],
              condition: { expression: 'true' },
            },
            {
              name: 'super-admin_v2',
              parentRoles: ['user'],
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).not.toThrow();
    });
  });

  describe('Parent Role Validation', () => {
    it('should accept wildcard "*" parent role', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'wildcard-policy' },
        spec: {
          definitions: [
            {
              name: 'authenticated',
              parentRoles: ['*'],
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).not.toThrow();
    });

    it('should accept prefix wildcard "prefix:*" parent role', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'prefix-wildcard-policy' },
        spec: {
          definitions: [
            {
              name: 'admin',
              parentRoles: ['admin:*'],
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).not.toThrow();
    });

    it('should reject invalid wildcard patterns', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'invalid-wildcard' },
        spec: {
          definitions: [
            {
              name: 'admin',
              parentRoles: ['admin*'], // Invalid - should be admin:*
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).toThrow('Invalid parent role pattern');
    });

    it('should reject wildcards in the middle of role names', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'invalid-wildcard' },
        spec: {
          definitions: [
            {
              name: 'admin',
              parentRoles: ['admin:*:role'], // Invalid - wildcard must be at end
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).toThrow('Invalid parent role pattern');
    });
  });

  describe('Circular Dependency Detection', () => {
    it('should detect simple circular dependency (A -> B -> A)', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'circular' },
        spec: {
          definitions: [
            {
              name: 'roleA',
              parentRoles: ['roleB'],
              condition: { expression: 'true' },
            },
            {
              name: 'roleB',
              parentRoles: ['roleA'],
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).toThrow(CircularDependencyError);
    });

    it('should detect complex circular dependency (A -> B -> C -> A)', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'circular' },
        spec: {
          definitions: [
            {
              name: 'roleA',
              parentRoles: ['roleB'],
              condition: { expression: 'true' },
            },
            {
              name: 'roleB',
              parentRoles: ['roleC'],
              condition: { expression: 'true' },
            },
            {
              name: 'roleC',
              parentRoles: ['roleA'],
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).toThrow(CircularDependencyError);
    });

    it('should detect self-referencing role', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'self-ref' },
        spec: {
          definitions: [
            {
              name: 'recursive',
              parentRoles: ['recursive'],
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).toThrow(CircularDependencyError);
    });

    it('should allow roles depending on base roles (not derived)', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'valid' },
        spec: {
          definitions: [
            {
              name: 'owner',
              parentRoles: ['user'], // 'user' is a base role
              condition: { expression: 'true' },
            },
            {
              name: 'admin',
              parentRoles: ['superuser'], // 'superuser' is a base role
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).not.toThrow();
    });

    it('should allow valid dependency chain without cycles', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'chain' },
        spec: {
          definitions: [
            {
              name: 'level1',
              parentRoles: ['user'],
              condition: { expression: 'true' },
            },
            {
              name: 'level2',
              parentRoles: ['level1'],
              condition: { expression: 'true' },
            },
            {
              name: 'level3',
              parentRoles: ['level2'],
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).not.toThrow();
    });

    it('should handle diamond dependency (A -> B, A -> C, B -> D, C -> D)', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'diamond' },
        spec: {
          definitions: [
            {
              name: 'roleD',
              parentRoles: ['user'],
              condition: { expression: 'true' },
            },
            {
              name: 'roleB',
              parentRoles: ['roleD'],
              condition: { expression: 'true' },
            },
            {
              name: 'roleC',
              parentRoles: ['roleD'],
              condition: { expression: 'true' },
            },
            {
              name: 'roleA',
              parentRoles: ['roleB', 'roleC'],
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).not.toThrow();
    });

    it('should handle multiple independent chains', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'multi-chain' },
        spec: {
          definitions: [
            // Chain 1
            {
              name: 'chain1_a',
              parentRoles: ['user'],
              condition: { expression: 'true' },
            },
            {
              name: 'chain1_b',
              parentRoles: ['chain1_a'],
              condition: { expression: 'true' },
            },
            // Chain 2
            {
              name: 'chain2_a',
              parentRoles: ['admin'],
              condition: { expression: 'true' },
            },
            {
              name: 'chain2_b',
              parentRoles: ['chain2_a'],
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).not.toThrow();
    });

    it('should detect cycle in one chain while others are valid', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'partial-cycle' },
        spec: {
          definitions: [
            // Valid chain
            {
              name: 'valid_a',
              parentRoles: ['user'],
              condition: { expression: 'true' },
            },
            {
              name: 'valid_b',
              parentRoles: ['valid_a'],
              condition: { expression: 'true' },
            },
            // Circular chain
            {
              name: 'circular_a',
              parentRoles: ['circular_b'],
              condition: { expression: 'true' },
            },
            {
              name: 'circular_b',
              parentRoles: ['circular_a'],
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy])).toThrow(CircularDependencyError);
    });
  });

  describe('Edge Cases', () => {
    it('should handle empty policies array', () => {
      expect(() => validator.validate([])).not.toThrow();
    });

    it('should handle policy with empty definitions', () => {
      const policy: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'empty' },
        spec: {
          definitions: [],
        },
      };

      expect(() => validator.validate([policy])).not.toThrow();
    });

    it('should handle multiple policies with no conflicts', () => {
      const policy1: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'policy1' },
        spec: {
          definitions: [
            {
              name: 'role1',
              parentRoles: ['user'],
              condition: { expression: 'true' },
            },
          ],
        },
      };

      const policy2: ValidatedDerivedRolesPolicy = {
        apiVersion: 'authz.engine/v1',
        kind: 'DerivedRoles',
        metadata: { name: 'policy2' },
        spec: {
          definitions: [
            {
              name: 'role2',
              parentRoles: ['admin'],
              condition: { expression: 'true' },
            },
          ],
        },
      };

      expect(() => validator.validate([policy1, policy2])).not.toThrow();
    });
  });
});

File: go-core/tests/derived_roles/validator_test.go
package derived_roles_test

import (
	"fmt"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	"github.com/authz-engine/go-core/internal/derived_roles"
	"github.com/authz-engine/go-core/pkg/types"
)

// TestSchemaValidation tests derived roles schema validation
func TestSchemaValidation(t *testing.T) {
	t.Run("should accept valid derived role", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		derivedRole := &types.DerivedRole{
			Name:        "owner",
			ParentRoles: []string{"user"},
			Condition:   "resource.attr.ownerId == principal.id",
		}

		err = validator.Validate(derivedRole)
		require.NoError(t, err)
	})

	t.Run("should reject nil derived role", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		err = validator.Validate(nil)
		require.Error(t, err)
		assert.Contains(t, err.Error(), "nil", "Should mention nil")
	})

	t.Run("should reject duplicate role names in collection", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		derivedRoles := []*types.DerivedRole{
			{Name: "owner", ParentRoles: []string{"user"}, Condition: "true"},
			{Name: "owner", ParentRoles: []string{"admin"}, Condition: "true"}, // Duplicate
		}

		err = validator.ValidateAll(derivedRoles)
		require.Error(t, err)
		assert.Contains(t, err.Error(), "duplicate", "Should mention duplicate")
		assert.Contains(t, err.Error(), "owner", "Should mention the duplicate role name")
	})

	t.Run("should accept role names with hyphens and underscores", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		validNames := []string{"owner-role", "admin_role", "super-admin_v2"}
		for _, name := range validNames {
			derivedRole := &types.DerivedRole{
				Name:        name,
				ParentRoles: []string{"user"},
				Condition:   "true",
			}

			err = validator.Validate(derivedRole)
			require.NoError(t, err, "Should accept role name: %s", name)
		}
	})

	t.Run("should reject empty role names", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		derivedRole := &types.DerivedRole{
			Name:        "",
			ParentRoles: []string{"user"},
			Condition:   "true",
		}

		err = validator.Validate(derivedRole)
		require.Error(t, err)
		assert.Contains(t, err.Error(), "name", "Should mention name")
	})
}

// TestParentRoleValidation tests parent role validation
func TestParentRoleValidation(t *testing.T) {
	t.Run("should accept valid parent roles", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		derivedRole := &types.DerivedRole{
			Name:        "owner",
			ParentRoles: []string{"user", "admin"},
			Condition:   "true",
		}

		err = validator.Validate(derivedRole)
		require.NoError(t, err)
	})

	t.Run("should reject empty parent roles list", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		derivedRole := &types.DerivedRole{
			Name:        "owner",
			ParentRoles: []string{},
			Condition:   "true",
		}

		err = validator.Validate(derivedRole)
		require.Error(t, err)
		assert.Contains(t, err.Error(), "parent", "Should mention parent roles")
	})

	t.Run("should reject self-reference in parent roles", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		derivedRole := &types.DerivedRole{
			Name:        "owner",
			ParentRoles: []string{"owner"}, // Self-reference
			Condition:   "true",
		}

		err = validator.Validate(derivedRole)
		require.Error(t, err)
		assert.Contains(t, err.Error(), "itself", "Should mention self-reference")
	})

	t.Run("should accept wildcard parent roles", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		wildcards := [][]string{
			{"*"},
			{"admin:*"},
			{"*:manager"},
		}

		for _, parentRoles := range wildcards {
			derivedRole := &types.DerivedRole{
				Name:        "test-role",
				ParentRoles: parentRoles,
				Condition:   "true",
			}

			err = validator.Validate(derivedRole)
			require.NoError(t, err, "Should accept wildcard: %v", parentRoles)
		}
	})
}

// TestConditionValidation tests CEL condition validation
func TestConditionValidation(t *testing.T) {
	t.Run("should accept valid CEL expression", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		derivedRole := &types.DerivedRole{
			Name:        "owner",
			ParentRoles: []string{"user"},
			Condition:   "resource.attr.ownerId == principal.id",
		}

		err = validator.Validate(derivedRole)
		require.NoError(t, err)
	})

	t.Run("should accept empty condition (always true)", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		derivedRole := &types.DerivedRole{
			Name:        "owner",
			ParentRoles: []string{"user"},
			Condition:   "",
		}

		err = validator.Validate(derivedRole)
		require.NoError(t, err)
	})

	t.Run("should reject invalid CEL syntax", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		derivedRole := &types.DerivedRole{
			Name:        "owner",
			ParentRoles: []string{"user"},
			Condition:   "invalid syntax ::::",
		}

		err = validator.Validate(derivedRole)
		require.Error(t, err)
		assert.Contains(t, err.Error(), "condition", "Should mention condition")
	})

	t.Run("should accept complex boolean expressions", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		complexConditions := []string{
			"principal.attr.age > 18 && principal.attr.verified == true",
			"resource.kind == 'document' || resource.kind == 'file'",
			"principal.id in resource.attr.owners",
			"resource.attr.tags.exists(t, t == 'public')",
		}

		for _, condition := range complexConditions {
			derivedRole := &types.DerivedRole{
				Name:        "test-role",
				ParentRoles: []string{"user"},
				Condition:   condition,
			}

			err = validator.Validate(derivedRole)
			require.NoError(t, err, "Should accept condition: %s", condition)
		}
	})
}

// TestCircularDependencyDetection tests circular dependency detection
func TestCircularDependencyDetection(t *testing.T) {
	t.Run("should detect simple circular dependency", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		derivedRoles := []*types.DerivedRole{
			{
				Name:        "role1",
				ParentRoles: []string{"role2"},
				Condition:   "true",
			},
			{
				Name:        "role2",
				ParentRoles: []string{"role1"},
				Condition:   "true",
			},
		}

		err = validator.ValidateAll(derivedRoles)
		require.Error(t, err)
		assert.Contains(t, err.Error(), "circular", "Should mention circular dependency")
	})

	t.Run("should detect transitive circular dependency", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		derivedRoles := []*types.DerivedRole{
			{
				Name:        "role1",
				ParentRoles: []string{"role2"},
				Condition:   "true",
			},
			{
				Name:        "role2",
				ParentRoles: []string{"role3"},
				Condition:   "true",
			},
			{
				Name:        "role3",
				ParentRoles: []string{"role1"}, // Circular: role1 -> role2 -> role3 -> role1
				Condition:   "true",
			},
		}

		err = validator.ValidateAll(derivedRoles)
		require.Error(t, err)
		assert.Contains(t, err.Error(), "circular", "Should detect transitive circular dependency")
	})

	t.Run("should accept acyclic dependency graph", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		derivedRoles := []*types.DerivedRole{
			{
				Name:        "level1",
				ParentRoles: []string{"user"},
				Condition:   "true",
			},
			{
				Name:        "level2",
				ParentRoles: []string{"level1"},
				Condition:   "true",
			},
			{
				Name:        "level3",
				ParentRoles: []string{"level2"},
				Condition:   "true",
			},
		}

		err = validator.ValidateAll(derivedRoles)
		require.NoError(t, err, "Should accept acyclic graph")
	})
}

// TestComplexValidationScenarios tests complex multi-aspect validation
func TestComplexValidationScenarios(t *testing.T) {
	t.Run("should validate large collection of derived roles", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		// Create 50 derived roles with various dependencies
		derivedRoles := make([]*types.DerivedRole, 50)
		for i := 0; i < 50; i++ {
			parentRoles := []string{"user"}
			if i > 0 {
				// Add dependency on previous role
				parentRoles = append(parentRoles, fmt.Sprintf("role-%d", i-1))
			}

			derivedRoles[i] = &types.DerivedRole{
				Name:        fmt.Sprintf("role-%d", i),
				ParentRoles: parentRoles,
				Condition:   "true",
			}
		}

		err = validator.ValidateAll(derivedRoles)
		require.NoError(t, err, "Should validate large collection efficiently")
	})

	t.Run("should validate complex dependency graph", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		derivedRoles := []*types.DerivedRole{
			{Name: "editor", ParentRoles: []string{"user"}, Condition: "true"},
			{Name: "owner", ParentRoles: []string{"editor"}, Condition: "true"},
			{Name: "admin", ParentRoles: []string{"owner"}, Condition: "true"},
			{Name: "moderator", ParentRoles: []string{"editor"}, Condition: "true"},
			{Name: "superadmin", ParentRoles: []string{"admin", "moderator"}, Condition: "true"},
		}

		err = validator.ValidateAll(derivedRoles)
		require.NoError(t, err, "Should validate complex dependency graph")
	})
}

// TestEdgeCases tests validation edge cases
func TestValidatorEdgeCases(t *testing.T) {
	t.Run("should handle empty derived roles list", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		err = validator.ValidateAll([]*types.DerivedRole{})
		require.NoError(t, err, "Empty list should be valid")
	})

	t.Run("should handle single derived role", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		derivedRoles := []*types.DerivedRole{
			{Name: "owner", ParentRoles: []string{"user"}, Condition: "true"},
		}

		err = validator.ValidateAll(derivedRoles)
		require.NoError(t, err)
	})

	t.Run("should validate role with multiple parent roles", func(t *testing.T) {
		validator, err := derived_roles.NewDerivedRolesValidator()
		require.NoError(t, err)

		derivedRole := &types.DerivedRole{
			Name:        "superuser",
			ParentRoles: []string{"admin", "editor", "moderator"},
			Condition:   "true",
		}

		err = validator.Validate(derivedRole)
		require.NoError(t, err)
	})
}
